pyelftool_parser/src/execute.py

```diff
@@ -0,0 +1,212 @@
+from capstone import *
+from capstone.x86 import *
+from debug import loginst,log, logcall, logerror, logstack
+class execute:
+    def __init__(self, register):
+        self.register = register
+        self.reg = register.reg
+        self.retflag = 0
+    def exe(self, inst, edge, vertexfrom):
+        ## -----------------------------------------------
+        (regs_read, regs_write) = inst.regs_access()
+        ##  catch the rsp reg in instruction.    
+        flagrsp = 0
+        src = 0
+        dst = 0
+        reg = []
+        readreg = []
+        writereg = []
+        flagimm = 0
+        flagmem = 0
+        imm = 0
+        disp = 0
+        self.retflag = 0
+        ############# decode stage.
+        loginst("instruction")
+        loginst(hex(inst.address), inst.mnemonic, inst.op_str)
+        if "ptr" in inst.op_str:    ## early exit for ptr, I do not handle the pointer to memory yet.
+            loginst(inst.address, inst.mnemonic, inst.op_str)
+            loginst("I do not handle ptr memory yet")
+            return 0
+        if len(inst.operands) >= 2:
+            if (inst.id == X86_INS_FXCH):
+                dst = inst.op_str
+                log(dst)
+            elif (inst.id == X86_INS_LEA):
+                src = inst.op_str.split(",")[1].replace("[","").replace("]","")
+                ## This eval might be buggy because I assume we have only bit 64 register in the ptr for LEA instruction.
+                src = eval(src,self.reg)
+                dst = inst.op_str.split(",")[0]
+            else:
+                src = inst.op_str.split(",")[1].replace(" ","")
+                dst = inst.op_str.split(",")[0]
+        elif len(inst.operands) == 1:
+            dst = inst.op_str
+        for i in inst.operands:
+            if i.type == X86_OP_REG:
+                reg.append(inst.reg_name(i.reg))
+            if i.type == X86_OP_IMM:
+                imm = i.imm
+                flagimm = 1
+            if i.type == X86_OP_MEM:
+                base = inst.reg_name(i.mem.base)
+                disp = i.mem.disp
+                flagmem = 1
+        for r in regs_read:  ## catch the implicity read register of stack
+            readreg.append(inst.reg_name(r))
+            if "rsp" in inst.reg_name(r):
+                flagrsp = 1
+        for r in regs_write: ## catch the implicity write register of stack
+            writereg.append(inst.reg_name(r))
+            if "rsp" in inst.reg_name(r):
+                flagrsp = 1
+        ############################################
+        ##------------------------------------------
+        ## execute stage.
+        if flagrsp:  ## if rsp is in the instruction
```

Comment:
Isn't there a part of the API that tells you if an instruction modifies the stack? Aren't those the only instructions you need to worry about?

Reply:
I use capstone's "(regs_read, regs_write) = inst.regs_access()". Check that is it reading/writing rsp? This regs_access can also specify the implication read/write rsp.